File: web/src/api/visual-mood.ts
import { z } from "zod";
import { digest, stableId } from "./trip-context";
import { canonicalHash, halfUp, MOOD_KEYS, MOOD_POLICY_SHA, ordered, same, without, nullableScore } from "./grounded-core";

export const moodObservationSchema = z.strictObject({ dimension: z.enum(MOOD_KEYS), state: z.enum(["OBSERVED", "UNKNOWN"]),
  level: z.number().int().min(0).max(4).nullable(), certainty: z.enum(["HIGH", "LOW"]) })
  .refine((row) => row.state === "UNKNOWN" ? row.level === null && row.certainty === "LOW" : row.level !== null && row.certainty === "HIGH");
export const moodCandidateSetSchema = z.strictObject({
  schema_version: z.literal("photo-mood-candidates.v1"), family: z.literal("photo-mood-v1"), mood_version: z.literal("visual-mood-v1"),
  authority_scope: z.literal("VISUAL_MOOD_ONLY"), job_id: digest, image_index: z.number().int().min(1).max(3), payload_sha256: digest,
  policy_sha256: z.literal(MOOD_POLICY_SHA), provider_id: z.string().min(1).max(100), analysis_kind: z.enum(["MODEL", "SYNTHETIC"]),
  model: z.literal("glm-5.3-flash").nullable(), candidates: z.array(z.strictObject({ candidate_id: digest, observation: moodObservationSchema })).length(8),
  candidate_set_sha256: digest,
}).refine((row) => same(row.candidates.map((candidate) => candidate.observation.dimension), MOOD_KEYS) &&
  (row.analysis_kind === "SYNTHETIC" ? row.model === null && row.candidates.every((candidate) => candidate.observation.state === "UNKNOWN")
    : row.model === "glm-5.3-flash" && !row.provider_id.toLowerCase().includes("synthetic")));
export const projectedMoodSchema = z.strictObject({ dimension: z.enum(MOOD_KEYS), value: nullableScore,
  distinct_images: z.number().int().min(0).max(3), candidate_ids: z.array(digest).refine(ordered) })
  .refine((row) => (row.value === null) === (row.distinct_images === 0) && row.candidate_ids.length === row.distinct_images);
export const moodChoiceSchema = z.strictObject({ candidate_id: digest, included: z.boolean() });
export const confirmedMoodSchema = z.strictObject({
  schema_version: z.literal("photo-mood-projection.v1"), family: z.literal("photo-mood-v1"), job_id: digest,
  preference_profile_id: stableId, policy_sha256: z.literal(MOOD_POLICY_SHA), draft_sha256: digest,
  candidate_set_sha256: z.array(digest).refine(ordered), choices: z.array(moodChoiceSchema).max(24),
  moods: z.array(projectedMoodSchema).length(8), receipt_id: digest,
}).refine((row) => same(row.moods.map((mood) => mood.dimension), MOOD_KEYS) && ordered(row.choices.map((choice) => choice.candidate_id)));
export const moodReviewSchema = z.strictObject({
  schema_version: z.literal("photo-mood-review.v1"), family: z.literal("photo-mood-v1"), job_id: digest,
  preference_profile_id: stableId, draft_sha256: digest, batches: z.array(moodCandidateSetSchema).max(3), confirmation: confirmedMoodSchema.nullable(),
});
export type MoodCandidateSet = z.infer<typeof moodCandidateSetSchema>;
export type ProjectedMood = z.infer<typeof projectedMoodSchema>;
export type ConfirmedMood = z.infer<typeof confirmedMoodSchema>;
export type MoodReview = z.infer<typeof moodReviewSchema>;

export async function validateMoodBatch(batch: MoodCandidateSet): Promise<boolean> {
  if (batch.candidate_set_sha256 !== await canonicalHash(without(batch, ["candidate_set_sha256"]))) return false;
  for (const row of batch.candidates) {
    if (row.candidate_id !== await canonicalHash({ job_id: batch.job_id, image: batch.payload_sha256,
      observation: row.observation, provider_id: batch.provider_id, policy_sha256: batch.policy_sha256 })) return false;
  }
  return true;
}

export function aggregateMoodBatches(batches: MoodCandidateSet[], selected?: string[]): ProjectedMood[] | null {
  const images = new Map<string, MoodCandidateSet>();
  for (const batch of batches) {
    const previous = images.get(batch.payload_sha256);
    if (previous && !same(previous.candidates, batch.candidates)) return null;
    images.set(batch.payload_sha256, batch);
  }
  const candidates = [...images.values()].flatMap((batch) => batch.candidates);
  if (selected && selected.some((id) => !candidates.some((candidate) => candidate.candidate_id === id))) return null;
  return MOOD_KEYS.map((dimension) => {
    const rows = candidates.filter((candidate) => candidate.observation.dimension === dimension && candidate.observation.state === "OBSERVED" &&
      (!selected || selected.includes(candidate.candidate_id)));
    return { dimension, value: rows.length ? halfUp(rows.reduce((sum, row) => sum + row.observation.level! * 25, 0), rows.length) : null,
      distinct_images: rows.length, candidate_ids: rows.map((row) => row.candidate_id).sort() };
  });
}

export async function validateMoodConfirmation(confirmation: ConfirmedMood, review?: MoodReview): Promise<boolean> {
  if (confirmation.receipt_id !== await canonicalHash(without(confirmation, ["receipt_id"]))) return false;
  if (!review) return true;
  const candidates = new Map(review.batches.flatMap((batch) => batch.candidates).map((candidate) => [candidate.candidate_id, candidate]));
  return confirmation.job_id === review.job_id && confirmation.preference_profile_id === review.preference_profile_id &&
    confirmation.draft_sha256 === review.draft_sha256 && same(confirmation.candidate_set_sha256, [...new Set(review.batches.map((batch) => batch.candidate_set_sha256))].sort()) &&
    confirmation.choices.every((choice) => candidates.has(choice.candidate_id) && (!choice.included || candidates.get(choice.candidate_id)!.observation.state === "OBSERVED")) &&
    same(confirmation.moods, aggregateMoodBatches(review.batches, confirmation.choices.filter((choice) => choice.included).map((choice) => choice.candidate_id)));
}

export async function validateMoodReview(review: MoodReview): Promise<boolean> {
  if (review.batches.some((batch) => batch.job_id !== review.job_id) ||
    !(await Promise.all(review.batches.map(validateMoodBatch))).every(Boolean)) return false;
  const draft = await canonicalHash({ schema_version: "photo-mood-draft.v1", job_id: review.job_id,
    preference_profile_id: review.preference_profile_id, policy_sha256: MOOD_POLICY_SHA,
    candidate_set_sha256: [...new Set(review.batches.map((batch) => batch.candidate_set_sha256))].sort() });
  return review.draft_sha256 === draft && (review.confirmation === null || await validateMoodConfirmation(review.confirmation, review));
}

File: backend/src/itda/contracts/visual_mood.py
"""Independent appearance observations; no factual tourism score authority."""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Literal, Self

from pydantic import Field, StrictBool, model_validator

from itda.contracts.base import Sha256, StableId, StrictContract
from itda.domain.canonical import canonical_sha256

VISUAL_MOOD_VERSION: Literal["visual-mood-v1"] = "visual-mood-v1"
PHOTO_MOOD_FAMILY: Literal["photo-mood-v1"] = "photo-mood-v1"


class VisualMoodDimension(StrEnum):
    GREENERY = "greenery"
    WATER = "water"
    OPEN_COMPOSITION = "open_composition"
    TRADITIONAL_APPEARANCE = "traditional_appearance"
    CONTEMPORARY_DESIGN = "contemporary_design"
    WARM_LIGHT = "warm_light"
    VIVID_COLOR = "vivid_color"
    NIGHT_LIGHTING = "night_lighting"


MOOD_LABELS = {
    VisualMoodDimension.GREENERY: "초록 식물이 보이는 풍경",
    VisualMoodDimension.WATER: "물이 보이는 풍경",
    VisualMoodDimension.OPEN_COMPOSITION: "시야가 트여 보이는 구도",
    VisualMoodDimension.TRADITIONAL_APPEARANCE: "전통적으로 보이는 외관",
    VisualMoodDimension.CONTEMPORARY_DESIGN: "현대적으로 보이는 디자인",
    VisualMoodDimension.WARM_LIGHT: "따뜻한 빛의 색감",
    VisualMoodDimension.VIVID_COLOR: "선명한 색감",
    VisualMoodDimension.NIGHT_LIGHTING: "밤 조명이 보이는 장면",
}
MOOD_RUBRIC = {
    "version": VISUAL_MOOD_VERSION,
    "levels": {
        "0": "직접 보이는 해당 시각 요소가 매우 약함",
        "1": "약함",
        "2": "보통",
        "3": "강함",
        "4": "장면을 지배함",
    },
    "unknown": "이미지에서 직접 판단할 수 없거나 확신이 낮으면 UNKNOWN/null/LOW",
    "scope": (
        "사진의 외관·빛·색·구도만 평가. 실제 혼잡, 인원 수, 복잡도, 운영, 시설, "
        "역사적 진위, 활동, 체류, OCR 금지."
    ),
    "dimensions": {dimension.value: label for dimension, label in MOOD_LABELS.items()},
}
MOOD_POLICY_SHA256 = canonical_sha256(MOOD_RUBRIC)


class MoodObservation(StrictContract):
    dimension: VisualMoodDimension
    state: Literal["OBSERVED", "UNKNOWN"]
    level: Annotated[int, Field(strict=True, ge=0, le=4)] | None
    certainty: Literal["HIGH", "LOW"]

    @model_validator(mode="after")
    def validate_observation(self) -> Self:
        if self.state == "UNKNOWN" and (self.level is not None or self.certainty != "LOW"):
            raise ValueError("unknown appearance requires null and low certainty")
        if self.state == "OBSERVED" and (self.level is None or self.certainty != "HIGH"):
            raise ValueError("observed appearance requires direct high-certainty support")
        return self


class MoodWireResponse(StrictContract):
    observations: Annotated[tuple[MoodObservation, ...], Field(min_length=8, max_length=8)]

    @model_validator(mode="after")
    def full_vocabulary(self) -> Self:
        if {row.dimension for row in self.observations} != set(VisualMoodDimension):
            raise ValueError("appearance response must cover the exact closed vocabulary")
        return self


class PhotoMoodCandidate(StrictContract):
    candidate_id: Sha256
    observation: MoodObservation


class PhotoMoodCandidateSet(StrictContract):
    schema_version: Literal["photo-mood-candidates.v1"] = "photo-mood-candidates.v1"
    family: Literal["photo-mood-v1"] = PHOTO_MOOD_FAMILY
    mood_version: Literal["visual-mood-v1"] = VISUAL_MOOD_VERSION
    authority_scope: Literal["VISUAL_MOOD_ONLY"] = "VISUAL_MOOD_ONLY"
    job_id: Sha256
    image_index: Annotated[int, Field(strict=True, ge=1, le=3)]
    payload_sha256: Sha256
    policy_sha256: Sha256
    provider_id: Annotated[str, Field(min_length=1, max_length=100)]
    analysis_kind: Literal["MODEL", "SYNTHETIC"]
    model: Literal["glm-5.3-flash"] | None
    candidates: Annotated[tuple[PhotoMoodCandidate, ...], Field(min_length=8, max_length=8)]
    candidate_set_sha256: Sha256

    @model_validator(mode="after")
    def validate_set(self) -> Self:
        if self.policy_sha256 != MOOD_POLICY_SHA256:
            raise ValueError("unknown appearance policy")
        if tuple(row.observation.dimension for row in self.candidates) != tuple(
            VisualMoodDimension
        ):
            raise ValueError("appearance candidates require canonical order")
        if self.analysis_kind == "SYNTHETIC":
            if self.model is not None or any(
                row.observation.state != "UNKNOWN" for row in self.candidates
            ):
                raise ValueError("synthetic mode has no visual inference authority")
        elif self.model != "glm-5.3-flash" or "synthetic" in self.provider_id.casefold():
            raise ValueError("appearance model provenance invalid")
        for row in self.candidates:
            expected = canonical_sha256(
                {
                    "job_id": self.job_id,
                    "image": self.payload_sha256,
                    "observation": row.observation.model_dump(mode="json"),
                    "provider_id": self.provider_id,
                    "policy_sha256": self.policy_sha256,
                }
            )
            if row.candidate_id != expected:
                raise ValueError("appearance candidate identity invalid")
        if self.candidate_set_sha256 != canonical_sha256(
            self.model_dump(mode="json", exclude={"candidate_set_sha256"})
        ):
            raise ValueError("appearance candidate digest invalid")
        return self


class ProjectedMood(StrictContract):
    dimension: VisualMoodDimension
    value: Annotated[int, Field(strict=True, ge=0, le=100)] | None
    distinct_images: Annotated[int, Field(strict=True, ge=0, le=3)]
    candidate_ids: tuple[Sha256, ...]

    @model_validator(mode="after")
    def validate_projection(self) -> Self:
        if (self.value is None) != (self.distinct_images == 0):
            raise ValueError("unknown appearance cannot be scored")
        if self.candidate_ids != tuple(sorted(set(self.candidate_ids))):
            raise ValueError("appearance references must be canonical")
        if self.distinct_images != len(self.candidate_ids):
            raise ValueError("appearance projection needs one candidate per distinct image")
        return self


class MoodChoice(StrictContract):
    candidate_id: Sha256
    included: StrictBool


class PhotoMoodConfirmRequest(StrictContract):
    draft_sha256: Sha256
    choices: Annotated[tuple[MoodChoice, ...], Field(max_length=24)]


class ConfirmedMoodProjection(StrictContract):
    schema_version: Literal["photo-mood-projection.v1"] = "photo-mood-projection.v1"
    family: Literal["photo-mood-v1"] = PHOTO_MOOD_FAMILY
    job_id: Sha256
    preference_profile_id: StableId
    policy_sha256: Sha256
    draft_sha256: Sha256
    candidate_set_sha256: tuple[Sha256, ...]
    choices: tuple[MoodChoice, ...]
    moods: Annotated[tuple[ProjectedMood, ...], Field(min_length=8, max_length=8)]
    receipt_id: Sha256

    @model_validator(mode="after")
    def validate_receipt(self) -> Self:
        if self.policy_sha256 != MOOD_POLICY_SHA256:
            raise ValueError("unknown mood projection policy")
        if tuple(row.dimension for row in self.moods) != tuple(VisualMoodDimension):
            raise ValueError("mood projection order invalid")
        if self.candidate_set_sha256 != tuple(sorted(set(self.candidate_set_sha256))):
            raise ValueError("mood batch hashes must be canonical")
        if tuple(row.candidate_id for row in self.choices) != tuple(
            sorted({row.candidate_id for row in self.choices})
        ):
            raise ValueError("mood choices must be canonical")
        if self.receipt_id != canonical_sha256(
            self.model_dump(mode="json", exclude={"receipt_id"})
        ):
            raise ValueError("mood receipt hash invalid")
        return self


class PhotoMoodReview(StrictContract):
    schema_version: Literal["photo-mood-review.v1"] = "photo-mood-review.v1"
    family: Literal["photo-mood-v1"] = PHOTO_MOOD_FAMILY
    job_id: Sha256
    preference_profile_id: StableId
    draft_sha256: Sha256
    batches: Annotated[tuple[PhotoMoodCandidateSet, ...], Field(max_length=3)]
    confirmation: ConfirmedMoodProjection | None = None
